src/Microsoft.Azure.SignalR.AspNet/Infrastructure/ServiceConnection.cs

```diff
@@ -153,5 +239,53 @@ private static string GetString(ReadOnlySequence<byte> buffer)
 
             return Encoding.UTF8.GetString(buffer.ToArray());
         }
+        
+        private sealed class ClientContext
+        {
+            public ClientContext()
+            {
+                CancellationToken = new CancellationTokenSource();
+
+                var channel = Channel.CreateUnbounded<ServiceMessage>();
+                Input = channel.Reader;
+                Output = channel.Writer;
+            }
+
+            public IServiceTransport Transport { get; set; }
+
+            public ChannelReader<ServiceMessage> Input { get; }
+            
+            public ChannelWriter<ServiceMessage> Output { get; }
+
+            public Task ApplicationTask { get; set; }
+
+            public CancellationTokenSource CancellationToken { get; set; }
+
+            public async Task DisposeAsync(bool closeGracefully = true)
+            {
+                try
+                {
+                    Output.Complete();
```

Comment:
TryComplete

Reply:
If try fails and won't throw exception, it will be blocked here pending waiting for task complete.

Reply:
This should never throw. When would it?

Reply:
Offline synced and TryComplete() returns false only when the channel is already completed. Will fix it.